File: frontend/src/lib/yjs/schema.ts
// Mirrors backend/app/services/collab/schema.py — change both sides together; the materializer reads exactly these keys.
//
// `day_notes` and item-level `notes` are `Y.Text` so concurrent edits merge instead of last-writer-wins. Pre-Phase-7f plans still hold plain strings in these slots; read paths tolerate both shapes and the first write upgrades the slot to a Y.Text.

export const ROOT_ITEMS = "items";
export const ROOT_DAY_NOTES = "day_notes";
export const ROOT_PLAN_META = "plan_meta";
export const ROOT_LIKES = "likes";
export const ROOT_RATINGS = "ratings";
export const ROOT_COMMENTS = "comments";

export const ITEM_FIELDS = [
  "id",
  "day_id",
  "destination_id",
  "place_id",
  "added_by",
  "item_type",
  "title",
  "notes",
  "location",
  "start_time",
  "end_time",
  "duration_minutes",
  "sort_key",
  "sort_order",
  "ai_data",
] as const;

export type ItemField = (typeof ITEM_FIELDS)[number];

export const PLAN_META_FIELDS = [
  "title",
  "description",
  "date_from",
  "date_to",
  "visibility",
  "cover_image_path",
  "cover_image_url",
] as const;

export type PlanMetaField = (typeof PLAN_META_FIELDS)[number];

export type PlanMetaPatch = Partial<Record<PlanMetaField, string | null>>;

// `id` is generated client-side at post time so the materializer can upsert by id without a DB round-trip to allocate one.
export const COMMENT_FIELDS = [
  "id",
  "plan_item_id",
  "parent_id",
  "author_id",
  "body",
  "created_at",
  "updated_at",
  "deleted_at",
] as const;

export type CommentField = (typeof COMMENT_FIELDS)[number];

// Awareness is ephemeral (not persisted, materializer ignores it). `editing` is set only on focus of a free-text surface — hover/idle viewing intentionally does not publish presence (too noisy in practice).
export interface AwarenessUser {
  id: string;
  displayName: string | null;
  username: string | null;
  avatarUrl: string | null;
}

export type EditingKind = "day_notes" | "item_notes" | "chat" | "item_comment";

export interface EditingTarget {
  kind: EditingKind;
  id: string;
}

export interface AwarenessState {
  user: AwarenessUser;
  editing: EditingTarget | null;
}

File: backend/app/services/collab/schema.py
"""Yjs document schema for Ajmo plans.

The same schema lives in `frontend/src/lib/yjs/schema.ts`. Whenever you change
either side, change both — the materializer assumes the field set defined
here is exactly what the frontend writes.

Scope: only the live-edited content lives in Yjs. The plan skeleton (days,
destinations) and hotels are REST-driven. Plan-meta is REST-driven *at rest*
but mirrored into a `plan_meta` Y.Map purely as a broadcast channel so
collaborators see title/date/cover changes without refreshing — the
materializer never reads or writes that root.

```
root (Y.Doc)
├── items:      Y.Map<Y.Array<Y.Map>>  # day_id → ordered list of items
├── day_notes:  Y.Map<string>          # day_id → notes text
└── plan_meta:  Y.Map<string|null>     # broadcast mirror of plan title/dates/cover/visibility
```
"""
from __future__ import annotations

ROOT_ITEMS = "items"
ROOT_DAY_NOTES = "day_notes"
ROOT_PLAN_META = "plan_meta"

ITEM_FIELDS = (
    "id",
    "day_id",
    "destination_id",
    "place_id",
    "added_by",
    "item_type",
    "title",
    "notes",
    "location",
    "start_time",
    "end_time",
    "duration_minutes",
    "sort_key",
    "sort_order",
    "ai_data",
)

PLAN_META_FIELDS = (
    "title",
    "description",
    "date_from",
    "date_to",
    "visibility",
    "cover_image_path",
    "cover_image_url",
)
